internal/pkg/policy/standalone.go

```diff
@@ -112,7 +113,43 @@ func (m *standAloneSelfMonitorT) check(ctx context.Context) {
 		message = fmt.Sprintf("Failed to request policies: %s", err)
 	}
 
+	remoteOutputErrorMap := m.bulker.GetRemoteOutputErrorMap()
+	hasError := false
+	remoteESPayload := make(map[string]interface{})
+	for key, value := range remoteOutputErrorMap {
+		if value != "" {
+			hasError = true
+			remoteESPayload[key] = value
+			break
+		}
+	}
+	if hasError {
+		state = client.UnitStateDegraded
+		message = fmt.Sprintf("Could not connect to remote ES output: %+v", remoteESPayload)
+		m.log.Debug().Msg(message)
+	} else {
+		bulkerMap := m.bulker.GetBulkerMap()
+		for outputName, outputBulker := range bulkerMap {
+			res, err := outputBulker.Client().Ping(outputBulker.Client().Ping.WithContext(ctx))
```

Comment:
This logic is not yet added to `self.go` (the monitor when fleet-server is run by agent)

There is a limitation if Fleet Server is restarted, the pinging will stop, because the `bulkerMap` is stored in memory, and it will be wiped out. 
Not sure what's the best way to solve this, should we store internal state on disk between restarts?

Reply:
added the remote es ping to the agent self monitor too in this commit: https://github.com/elastic/fleet-server/pull/3051/commits/8797823a149f773f73dae4ce4727168005a94a98

Reply:
On thinking about this more it's not really an issue that `bulkerMap` is cleared on FS restart, because the Prepare output logic is called on every agent checkin [here](https://github.com/elastic/fleet-server/blob/main/internal/pkg/api/handleCheckin.go#L753), so the pinging is going to restart as soon as an agent checks in that uses the remote output.

One question, where is the `message` field reported in standalone monitor? I'm seeing the status in `/api/status` call, but not the message.

Reply:
I'm also trying to test the agent based fleet-server self monitor, but I don't see fleet-server going to degraded state or the error logs from the self monitor, when I use an invalid remote es host.
I do see the error log that couldn't create API key, not sure what I'm doing wrong.
```
Could not create API key in remote ES
dial tcp 192.168.64.1:9202: connect: connection refused
```

Reply:
I noticed that the self monitor breaks the loop when it reaches healthy state: https://github.com/elastic/fleet-server/blob/772a5bf09e122b577696afe11256ef2f1ef8f2a3/internal/pkg/policy/self.go#L128-L130

If I change that to the loop running, is there any impact to consider? @blakerouse @cmacknz @michel-laterman

Reply:
I tested fleet-server with agent by removing the `break LOOP`, and I see fleet-server going to degraded state on the `/api/status` API. 
The fleet-server agent though doesn't change, it stays on healthy, and the component healthy doesn't change either. Is this implemented on Agent side to pick up degraded/unhealthy state from fleet-server?

Reply:
I would avoid saving remote output information locally, i would rather see it as part of a fleet index if required.
Right now we would have orphaned API keys if the output is removed from a policy and fleet-server restarts before the checkin requests (where the API keys would be invalidated) are processed. I think this may occur if the policy that the agent managing fleet-server is running has a remote output that is removed.

I think not checking a remote cluster until an agent checks in is OK.

I think breaking the self-monitor after it reaches the healthy state is to help/simplify the elastic-agent/fleet-server bootstrap process.
iirc the agent needs fleet to be degraded so it will send config, then healthy to checkin. (@cmacknz @blakerouse can one of you confirm?)
If we report degraded when a remote ES cluster is down/misconfigured the agent managing the fleet server may never go online. Or it may try to restart the fleet-server then go offline?

I think we need to consider this behaviour, and what it means for the fleet-server to be healthy from the agent's pov as well as a platform pov. This may mean adding a new endpoint so that `/api/status` can be the ready endpoint and another will be for full output health.